Source/Tools/LinuxEmulation/LinuxSyscalls/Seccomp/BPFEmitter.cpp

```diff
@@ -0,0 +1,378 @@
+      EMIT_INST(cmp(ARMEmitter::Size::i32Bit, REG_A, CompareSrcReg));
+    } else if (Op == BPF_JSET) {
+      CompareResultOp = ARMEmitter::Condition::CC_NE;
+      EMIT_INST(tst(ARMEmitter::Size::i32Bit, REG_A, CompareSrcReg));
+    }
+
+    fextl::unordered_map<uint32_t, ARMEmitter::ForwardLabel>::iterator TargetTrueLabel {};
+    fextl::unordered_map<uint32_t, ARMEmitter::ForwardLabel>::iterator TargetFalseLabel {};
+
+    if constexpr (!CalculateSize) {
+      TargetTrueLabel = JumpLabels.try_emplace(TargetTrue, ARMEmitter::ForwardLabel {}).first;
+      TargetFalseLabel = JumpLabels.try_emplace(TargetFalse, ARMEmitter::ForwardLabel {}).first;
+    }
+
+    EMIT_INST(b(CompareResultOp, &TargetTrueLabel->second));
+    EMIT_INST(b(&TargetFalseLabel->second));
+    break;
+  }
+  default: RETURN_ERROR(-EINVAL); // Unknown jump type
+  }
+
+  RETURN_SUCCESS();
+}
+
+template<bool CalculateSize>
+uint64_t BPFEmitter::HandleRet(uint32_t BPFIP, const sock_filter* Inst) {
+  [[maybe_unused]] size_t OpSize {};
+  const auto RValSrc = BPF_RVAL(Inst->code);
+  switch (RValSrc) {
+  case BPF_K: {
+    auto Const = ConstPool.try_emplace(Inst->k, ARMEmitter::ForwardLabel {});
+    EMIT_INST(ldr(ARMEmitter::WReg::w0, &Const.first->second));
+    break;
+  }
+  case BPF_X: EMIT_INST(mov(ARMEmitter::WReg::w0, REG_X)); break;
+  case BPF_A:
+    // w0 is already REG_A
+    static_assert(REG_A == ARMEmitter::WReg::w0, "This is expected to be the same");
+    break;
+  }
+
+  EMIT_INST(ret());
+
+  RETURN_SUCCESS();
+}
+
+template<bool CalculateSize>
+uint64_t BPFEmitter::HandleMisc(uint32_t BPFIP, const sock_filter* Inst) {
+  [[maybe_unused]] size_t OpSize {};
+  const auto MiscOp = BPF_MISCOP(Inst->code);
+  switch (MiscOp) {
+  case BPF_TAX: EMIT_INST(mov(REG_X, REG_A)); break;
+  case BPF_TXA: EMIT_INST(mov(REG_A, REG_X)); break;
+  default: RETURN_ERROR(-EINVAL) // Unsupported misc operation.
+  }
+
+  RETURN_SUCCESS();
+}
+
+template<bool CalculateSize, class Pred>
+uint64_t BPFEmitter::HandleEmission(uint32_t flags, const sock_fprog* prog) {
+  constexpr Pred PredFunc;
+  uint64_t CalculatedSize {};
+
+  for (uint32_t i = 0; i < prog->len; ++i) {
+    if constexpr (!CalculateSize) {
+      auto jump_label = JumpLabels.find(i);
+      if (jump_label != JumpLabels.end()) {
+        Bind(&jump_label->second);
+      }
+    }
+
+    bool HadError {};
+    uint64_t Result {};
+
+    const sock_filter* Inst = &prog->filter[i];
+    const uint16_t Code = Inst->code;
+    const uint16_t Class = BPF_CLASS(Code);
+    switch (Class) {
+    case BPF_LD:
+    case BPF_LDX: {
+      Result = HandleLoad<CalculateSize>(i, Inst);
+      HadError = PredFunc(Result);
+      break;
+    }
+    case BPF_ST:
+    case BPF_STX: {
+      Result = HandleStore<CalculateSize>(i, Inst);
+      HadError = PredFunc(Result);
+      break;
+    }
+    case BPF_ALU: {
+      Result = HandleALU<CalculateSize>(i, Inst);
+      HadError = PredFunc(Result);
```

Comment:
Pop this line out of the switch. Also, does this need a default?

Reply:
Popped out. default handler switched to FEX_UNREACHABLE since we handle all instruction classes.